pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,320 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	networkinginformers "k8s.io/client-go/informers/networking/v1alpha1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/events"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+	"k8s.io/utils/clock"
+	netutils "k8s.io/utils/net"
+)
+
+// Repair is a controller loop that periodically examines all service ClusterIP allocations
+// and logs any errors, and then sets the compacted and accurate list of all allocated IPs.
+//
+// Handles:
+// * Duplicate ClusterIP assignments caused by operator action or undetected race conditions
+// * ClusterIPs that do not match the currently configured range
+// * Allocations to services that were not actually created due to a crash or powerloss
+// * Migrates old versions of Kubernetes services into the atomic ipallocator model automatically
+// * ClusterIP out of range cused by a change on the serviceCIDR configured
+//
+// Can be run at infrequent intervals, and is best performed on startup of the master.
+// Is level driven and idempotent - all valid ClusterIPs will be updated into the ipallocator
+// at the end of a single execution loop if no race is encountered.
+//
+// There is a one-to-one relation between ClusterIPs --- IPAddresses
+//
+// Service_X [ClusterIP_X]  <------>  IPAddress_X [Ref:Service_X]   ok
+// Service_Y [ClusterIP_Y]  <------>  IPAddress_Y [Ref:Service_A]   bad, wrong reference
+//                          <------>  IPAddress_Z [Ref:Service_B]   bad, orphan IPAddress
+// Service_Z [ClusterIP_Z]  <------>  								bad, missing IPAddress
+
+type RepairIPAddress struct {
+	client   kubernetes.Interface
+	interval time.Duration
+
+	networkByFamily map[v1.IPFamily]*net.IPNet // networks we operate on, by their family
```

Comment:
you should use `netutils.IPFamily` here...
(this may require bumping `k8s.io/utils`)